scvi/dataset/dataset.py

```diff
@@ -1409,6 +1412,11 @@ def _highly_variable_genes(
 
         logger.info("extracting highly variable genes using {} flavor".format(flavor))
 
+        if flavor == "poisson_zeros":
+            var = poisson_gene_selection(self.X, n_top_genes, **highly_var_genes_kwargs)
+
+            return var
+
```

Comment:
It makes sense to me to put this below seurat v3

Reply:
We can then let line 1426 `counts=self.X.copy()` be `counts=sp_sparse.csr_matrix(self.X)` (maybe need a copy if it's already sparse?)

and then your new function can use the adata with potential batch information.

I think all the functions will be faster using sparse matrices.

Reply:
though maybe it's overkill to have your new function take an anndata object? the seurat v3 one really just needs X and the batch info. I guess it only really depends if you think people will want to access this function outside of `subsample_genes`? If not maybe the anndata is fine for now. We have someone coming to work on the codebase and he can refactor this kind of stuff.

Reply:
then your function doesn't need to return anything and the return value of `adata.var` would be consistent with the other functions.

Reply:
> though maybe it's overkill to have your new function take an anndata object? the seurat v3 one really just needs X and the batch info. I guess it only really depends if you think people will want to access this function outside of `subsample_genes`? If not maybe the anndata is fine for now. We have someone coming to work on the codebase and he can refactor this kind of stuff.

I was thinking the overhead of creating the anndata wouldn't be worth it. Though it's probably very minor. I implemented it like this first and used it outside of scVI. But then I figured since scVI requires PyTorch it would fit well as an internal function, and I figured it wouldn't need more than the X.